### PR TITLE
Add Active Directory support

### PR DESCRIPTION
**Pull request by Daniel-Trevitz**
For virtio-fs to work with an Active Directory connected system, we can't use NetUserGetInfo.
Instead, use WTSQueryUserToken to get a token for the security context of the logged-in user,
 and then query the user and group SID. Use the SID to get the UID using WinFSP's
 preexisting mapping, which for standard users is just the previously used 0x30000+RID.

Signed-off-by: Daniel Trevitz <daniel.trevitz@wika.com>

### CODE REVIEW

**Comment by Daniel-Trevitz** [comment]
See #660 and #790,  And pr #792

**Comment by Daniel-Trevitz** [comment]
New commit added. Address issues raised in PR #792.
Still have not worked on running unit tests. 

I have run this on both my vanilla win10 install and my AD connected install for both as a service and from the console. 

I have /not/ attempted session switching at all ... which I can only guess is switching users?

I do think there is something weird afoot, and /very/ probably AD related. 
The service start time is /drastically/ longer when I have an internet connection for the AD connected user.
That said, it's a one time price and the end result works so, meh.

**Comment by Daniel-Trevitz** [comment]
Winfsp releases their tests here: https://github.com/winfsp/winfsp/releases

**Comment by YanVugenfirer** [comment]
@Daniel-Trevitz 
"which I can only guess is switching users?"

Yes. And on Server SKUs connecting simultaneously with several users using terminal services.
"Advanced" test, would be testing access from the service without logged in users.

**Review by YanVugenfirer — APPROVED** [review]

File: viofs/svc/virtiofs.cpp
Comment (viktor-prutyanov) [inline viofs/svc/virtiofs.cpp:516]:
```suggestion
    Info = HeapAlloc(GetProcessHeap(), 0, Size);
    if (NULL == Info)
    {
        Result = STATUS_INSUFFICIENT_RESOURCES;
        goto exit;
    }
```

File: viofs/svc/virtiofs.cpp
Comment (viktor-prutyanov) [inline viofs/svc/virtiofs.cpp:522]:
```suggestion
    if (!GetTokenInformation(Token, TokenInformationClass, Info, Size, &Size))
    {
        Result = FspNtStatusFromWin32(GetLastError());
        goto exit;
    }
```

File: viofs/svc/virtiofs.cpp
Comment (viktor-prutyanov) [inline viofs/svc/virtiofs.cpp:532]:
```suggestion
    *PInfo = Info;
    Result = STATUS_SUCCESS;

exit:
    if (!NT_SUCCESS(Result))
        SafeHeapFree(Info);

    return Result;
}
```

File: viofs/svc/virtiofs.cpp
Comment (viktor-prutyanov) [inline viofs/svc/virtiofs.cpp:551]:
```suggestion
    if (VirtFs->IsRunningAsLocalSystem)
    {
        if (!WTSQueryUserToken(SessionId, &Token))
        {
            VirtFs->LocalUid = 0;
            VirtFs->LocalGid = 0;
            DBG("Failed to open a process token as LocalSystem; Error=%d", GetLastError());
            return;
        }
    }
```

File: viofs/svc/virtiofs.cpp
Comment (viktor-prutyanov) [inline viofs/svc/virtiofs.cpp:560]:
```suggestion
    Result = FspToolGetTokenInfo(Token, TokenUser, (PVOID *)&Uinfo);
```

File: viofs/svc/virtiofs.cpp
Comment (viktor-prutyanov) [inline viofs/svc/virtiofs.cpp:564]:
```suggestion
    Result = FspToolGetTokenInfo(Token, TokenPrimaryGroup, (PVOID *)&Ginfo);
```

File: viofs/svc/virtiofs.cpp
Comment (viktor-prutyanov) [inline viofs/svc/virtiofs.cpp:571]:
```suggestion
    DBG("Local UID=%u, local GID=%u", VirtFs->LocalUid, VirtFs->LocalGid);
```

File: viofs/svc/virtiofs.cpp
Comment (viktor-prutyanov) [inline viofs/svc/virtiofs.cpp:2593]:
```suggestion
    HANDLE Token{ 0 };
```

File: viofs/svc/virtiofs.cpp
Comment (viktor-prutyanov) [inline viofs/svc/virtiofs.cpp:2619]:
```suggestion
    // Will fail if we are not the LocalSystem user.
    IsRunningAsLocalSystem = WTSQueryUserToken(SessionId, &Token);

    if (IsRunningAsLocalSystem)
    {
        CloseHandle(Token);
    }
    else
    {
        DBG("The service %s was not run as the LocalSytem user", FS_SERVICE_NAME);
    }
```

**Review by viktor-prutyanov — COMMENTED** [review]
I have tested the PR as a console program and as a Windows service and it seems to work OK.
But I there are some codestyle fixes to apply before merge.

**Comment by Daniel-Trevitz** [comment]
> I have tested the PR as a console program and as a Windows service and it seems to work OK. But I there are some codestyle fixes to apply before merge.

Thanks; I was going to merge them from github, but it looks like it wants me to do a separate commit for each??
Since I have a subseqent topic, I'll figure out why the formatting is wrong here but right in the editor.

**Comment by Daniel-Trevitz** [comment]
@viktor-prutyanov; I see; notepad++ was using tabs. bleh, gross. Changes pushed.

**Comment by viktor-prutyanov** [comment]
@Daniel-Trevitz
I suggest you to fetch PR back to your local repo (`git fetch origin pull/795/head:pr795; get checkout pr795`) and squash all 3 commits together, because presence of "commit to fix codestyle" makes Git history dirty and hard to understand.

**Comment by viktor-prutyanov** [comment]
> I found the winfsp test suite, but I could not get it to pass. Would you be able to run that, and paste the command (and path you run it from) here with the outputs? Depending on the various combinations (including the one from [our winfsp-tests page](https://github.com/virtio-win/kvm-guest-drivers-windows/wiki/VirtIO-FS:-Shared-file-system#winfsp-tests) I can get various parts to pass or fail...

Yes, not all of the WinFsp tests passes at the moment on the FS. For now the working set is the following:
```
Z:
cd Z:\
"C:\Program Files (x86)\winfsp\bin\winfsp-tests-x64.exe" --resilient --external -reparse* -stream* -getfileinfo_test -create_fileattr_test -create_readonlydir_test -create_allocation_test -create_notraverse_test -create_backup_test -create_restore_test -create_share_test -getfileattr_test -getfileinfo_name_test -setfileinfo_test -setsecurity_test -querydir_test
```

**Comment by Daniel-Trevitz** [comment]
> 
> Yes, not all of the WinFsp tests passes at the moment on the FS. For now the working set is the following:
> 
@viktor-prutyanov, Okay. I was using it right. which is annoying:

```
Z:\>"C:\Program Files (x86)\winfsp\bin\winfsp-tests-x64.exe" --resilient --external -reparse* -stream* -getfileinfo_test -create_fileattr_test -create_readonlydir_test -create_allocation_test -create_notraverse_test -create_backup_test -create_restore_test -create_share_test -getfileattr_test -getfileinfo_name_test -setfileinfo_test -setsecurity_test -querydir_test
create_test............................ KO
    ASSERT(INVALID_HANDLE_VALUE != Handle) failed at create-test.c:43:create_dotest (err=c0000022:5)
```

**Comment by viktor-prutyanov** [comment]
Also, to make test run the directory should be clean from previous run artifacts.

**Comment by Daniel-Trevitz** [comment]
> Also, to make test run the directory should be clean from previous run artifacts.

Okay. I was not doing this, but more importantly I did not actually have write access again. Probably was using an old version, though I am not convinced it wouldn't just lose permission...

Now when running I get:
```
Z:\test>"C:\Program Files (x86)\winfsp\bin\winfsp-tests-x64.exe" --resilient --external -reparse* -stream* -getfileinfo_test -create_fileattr_test -create_readonlydir_test -create_allocation_test -create_notraverse_test -create_backup_test -create_restore_test -create_share_test -getfileattr_test -getfileinfo_name_test -setfileinfo_test -setsecurity_test -querydir_test
create_test............................ OK 0.02s
create_related_test.................... OK 0.01s
create_sd_test......................... OK 0.01s
create_curdir_test..................... OK 0.00s
create_namelen_test.................... OK 0.01s
delete_test............................ OK 0.01s
delete_access_test..................... OK 0.01s
delete_pending_test.................... OK 0.00s
delete_mmap_test....................... OK 0.00s
delete_standby_test.................... OK 0.07s
delete_ex_test......................... OK 0.00s
rename_test............................ KO
    ASSERT(ERROR_ALREADY_EXISTS == GetLastError()) failed at info-test.c:1189:rename_dotest (err=c0000022:5)
```

**Comment by Daniel-Trevitz** [comment]
Just looking through (info-test.c)[https://github.com/winfsp/winfsp/blob/master/tst/winfsp-tests/info-test.c] and it looks like line 1189 ... but I don't understand why that fails, because after the test is done there is a file1

~/test$ ls -R
.:
dir1  dir2

./dir1:
file0  file1

./dir2:

**Comment by Daniel-Trevitz** [comment]
> Just looking through (info-test.c)[https://github.com/winfsp/winfsp/blob/master/tst/winfsp-tests/info-test.c] and it looks like line 1189 ... but I don't understand why that fails, because after the test is done there is a file1

Oh, that's why. It fails differently every time I run it.

**Comment by YanVugenfirer** [comment]
Let's start with if it is a regression from master branch or not.

**Comment by viktor-prutyanov** [comment]
> Let's start with if it is a regression from master branch or not.

I don't see such a problem when running on a system without AD. I think we can accept changes which add some AD support and don't break non-AD setup even if they don't pass all of WinFsp in AD case.

**Comment by YanVugenfirer** [comment]
> > Let's start with if it is a regression from master branch or not.
> 
> I don't see such a problem when running on a system without AD. I think we can accept changes which add some AD support and don't break non-AD setup even if they don't pass all of WinFsp in AD case.

Agree.

**Comment by Daniel-Trevitz** [comment]
Squashed with `git rebase -i HEAD~4` followed by `git push origin +Add-AD`. +1 for new skill unlocked...

**Review by YanVugenfirer — APPROVED** [review]

**Comment by YanVugenfirer** [comment]
> I have tested the PR as a console program and as a Windows service and it seems to work OK.
> But I there are some codestyle fixes to apply before merge.
@viktor-prutyanov 
Github is blocking merge due to your review. For some reason it doesn't mark the review as resolved. Please take a look≥

In any case, waiting for build to finish.

File: viofs/svc/virtiofs.cpp
Comment (viktor-prutyanov) [inline viofs/svc/virtiofs.cpp:133]:
```suggestion
    BOOL IsRunningAsLocalSystem{ false };
```

File: viofs/svc/virtiofs.cpp
Comment (viktor-prutyanov) [inline viofs/svc/virtiofs.cpp:532]: This function still has 3 spaces as indentation width

File: viofs/svc/virtiofs.cpp
Comment (viktor-prutyanov) [inline viofs/svc/virtiofs.cpp:544]:
```suggestion
        if (!WTSQueryUserToken(SessionId, &Token))
```

**Review by viktor-prutyanov — CHANGES_REQUESTED** [review]

File: viofs/svc/virtiofs.cpp
Comment (Daniel-Trevitz) [reply viofs/svc/virtiofs.cpp:532]: Ah, the function I copied from winfsp. The horror. Only think worse is 2 spaces ;) On it.

Comment (viktor-prutyanov) [reply viofs/svc/virtiofs.cpp:532]: @YanVugenfirer I haven't marked as resolved mostly because of this

**Review by viktor-prutyanov — APPROVED** [review]

**Comment by Daniel-Trevitz** [comment]
Once the build server decides the commit is okay I'll push the request, but I think it's hung.

**Comment by YanVugenfirer** [comment]
The build also runs SDV (static driver verifier). But, yeah - it looks too much time. @kostyanf14 can you please take a look?

**Comment by Daniel-Trevitz** [comment]
Awesome! Green check marks across the board... but I don't see any button to push?
I do see a message saying only those with write access can merge pull requests.

**Comment by YanVugenfirer** [comment]
@Daniel-Trevitz merged.

Thanks a lot!

**Comment by jorins** [comment]
Thanks a bunch!! :heart: This one has pestered me for quite a while, glad to see it resolved

**Pull request merged**
